Lib/glob.py

```diff
@@ -310,3 +157,229 @@ def translate(pat, *, recursive=False, include_hidden=False, seps=None):
+    """
+    is_special = part in _special_parts
+    while parts and magic_check.search(parts[0]) is None:
+        # Consume next non-wildcard component (speeds up joining).
+        if parts[0] not in _special_parts:
+            is_special = False
+        part += os.path.sep + parts[0]
+        parts = parts[1:]
+
+    select_next = _selector(parts, recursive, include_hidden)
+
+    def select_literal(path, rel_path, dir_fd, exists):
+        path = _add_trailing_slash(path) + part
+        rel_path = _add_trailing_slash(rel_path) + part
+        yield from select_next(path, rel_path, dir_fd, exists and is_special)
+    return select_literal
+
+
+def _wildcard_selector(part, parts, recursive, include_hidden):
+    """Returns a function that selects direct children of a given path,
+    filtering by pattern.
+    """
+    if include_hidden and part == '*':
+        match = None  # Skip generating a pattern that would match all inputs.
+    else:
+        match = _compile_pattern(part, recursive, include_hidden)
+
+    if parts:
+        select_next = _selector(parts, recursive, include_hidden)
+        def select_wildcard(path, rel_path, dir_fd, exists):
+            close_fd = False
+            try:
+                arg, fd, close_fd = _open_dir(path, rel_path, dir_fd)
+                if fd is not None:
+                    prefix = _add_trailing_slash(path)
+                with os.scandir(arg) as scandir_it:
+                    entries = list(scandir_it)
+                for entry in entries:
+                    if match is None or match(entry.name):
+                        try:
+                            if entry.is_dir():
+                                entry_path = entry.path
+                                if fd is not None:
+                                    entry_path = prefix + entry_path
+                                yield from select_next(entry_path, entry.name, fd, True)
+                        except OSError:
+                            pass
+            except OSError:
+                pass
+            finally:
+                if close_fd:
+                    os.close(fd)
+
+    else:
+        def select_wildcard(path, rel_path, dir_fd, exists):
+            close_fd = False
+            try:
+                arg, fd, close_fd = _open_dir(path, rel_path, dir_fd)
+                prefix = _add_trailing_slash(path)
+                for name in os.listdir(arg):
+                    if match is None or match(name):
+                        yield prefix + name
+            except OSError:
+                pass
+            finally:
+                if close_fd:
+                    os.close(fd)
+
+    return select_wildcard
+
+
+def _recursive_selector(part, parts, recursive, include_hidden):
+    """Returns a function that selects a given path and all its children,
+    recursively, filtering by pattern.
+    """
+    while parts and parts[0] == '**':
+        parts = parts[1:]
```

Comment:
Please, use `pop(0)` or `list(dropwhile(lambda x: x == '**', parts))` to avoid creation of the new list on each iteration.

Reply:
It needs to be hashable because `_selector` is wrapped in an `lru_cache`! Otherwise I agree a list might be better.

Reply:
As you mentioned, it's not a part that require particular optimization, so it was more a nitpick :)